lib/solarwinds_apm/patch/tag_sql/sw_mysql2_patch.rb

```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+require_relative 'annotate_traceparent'
+
+module SolarWindsAPM
+  module Patch
+    module TagSql
+      module SWOMysql2Patch
+        def query(sql, options = {})
+          traceparent = AnnotateTraceparent.generate_traceparent
+          annotated_sql = ''
+
+          if traceparent.empty?
+            annotated_sql = sql
+          else
+            annotated_traceparent = "traceparent='#{AnnotateTraceparent.generate_traceparent}'"
+
+            current_span = ::OpenTelemetry::Trace.current_span
+            attributes_dup = current_span.attributes.dup
+            attributes_dup['sw.query_tag'] = "/*#{annotated_traceparent}*/"
+            current_span.instance_variable_set(:@attributes, attributes_dup.freeze)
+
+            annotated_sql = "#{sql} /*#{annotated_traceparent}*/"
+          end
+
+          super(annotated_sql, options)
+        end
+      end
+    end
+  end
+end
+
+# need to prepend before mysql2 instrumentation
+# so they will call this query function before
+# reach to original query function
+Mysql2::Client.prepend(SolarWindsAPM::Patch::TagSql::SWOMysql2Patch) if defined?(Mysql2::Client)
```

Comment:
Really glad you're point out this dependency. How is it being ensured that our patch is before the upstream instrumentation?

Reply:
upstream instrumentation will prepend the original function after `::OpenTelemetry::SDK.configure` is called (this step install all instrumentation). We just need to require this patch before sdk configure ([here](https://github.com/solarwinds/apm-ruby/pull/165/files#diff-3f9c82a89b0ea4082c8b5cffdd87df712142dda3e104a87ada619d2768fac4cdR115))

Reply:
Ahhh i see, very cool--can the comment be updated to be what you pointed out to me? And maybe it's worth noting it in the comment for [here](https://github.com/solarwinds/apm-ruby/pull/165/files#diff-3f9c82a89b0ea4082c8b5cffdd87df712142dda3e104a87ada619d2768fac4cdR115) too why it has to happen before `::OpenTelemetry::SDK.configure` is called.

Reply:
thinking more about this now as i look at the [upstream instrumentation](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/patches/client.rb)--its patch of `query` is creating the db client span, but is our patch of `query` actually executed before that? meaning at the time our patch runs, what is the span in scope?  from the example traces you showed me it seemed the correct span context was inserted but i'm wondering why it (and the simple `current_span.add_attribute()` works... maybe huddle is better to explain to me.

Reply:
Comments updated

Reply:
Gotcha, thanks for the clarification.